returns/contrib/pytest/plugin.py

```diff
@@ -166,65 +118,107 @@ def _trace_function(
+        (
+            'returns_lawful: all tests under `check_all_laws` ' +
+            'is marked this way, ' +
+            'use `-m "not returns_lawful"` to skip them.'
+        ),
+    )
+
+
+@pytest.fixture()
+def returns() -> Generator[ReturnsAsserts, None, None]:
+    """Returns class with helpers assertions to check containers."""
+    with _spy_error_handling() as errors_handled:
+        yield ReturnsAsserts(errors_handled)
+
+
+@contextmanager
+def _spy_error_handling() -> Generator[_ErrorsHandled, None, None]:
+    """Track error handling of containers."""
+    errs: _ErrorsHandled = {}
+    with ExitStack() as cleanup:
+        for container in _containers_to_patch():
+            for method, patch in _ERROR_HANDLING_PATCHERS.items():
+                cleanup.enter_context(mock.patch.object(
+                    container,
+                    method,
+                    patch(getattr(container, method), errs=errs),
+                ))
+        yield errs
+
+
+# delayed imports are needed to prevent messing up coverage
+def _containers_to_patch() -> tuple:
+    from returns.context import (
+        RequiresContextFutureResult,
+        RequiresContextIOResult,
+        RequiresContextResult,
+    )
+    from returns.future import FutureResult
+    from returns.io import IOFailure, IOSuccess
+    from returns.result import Failure, Success
+
+    return (
+        Success,
+        Failure,
+        IOSuccess,
+        IOFailure,
+        RequiresContextResult,
+        RequiresContextIOResult,
+        RequiresContextFutureResult,
+        FutureResult,
+    )
+
+
+def _patched_error_handler(
+    original: _FunctionType, errs: _ErrorsHandled,
+) -> _FunctionType:
+    if inspect.iscoroutinefunction(original):
+        async def wrapper(self, *args, **kwargs):
+            original_result = await original(self, *args, **kwargs)
+            errs[id(original_result)] = original_result
+            return original_result
+    else:
+        def wrapper(self, *args, **kwargs):
+            original_result = original(self, *args, **kwargs)
+            errs[id(original_result)] = original_result
+            return original_result
+    return wraps(original)(wrapper)  # type: ignore
+
+
+def _patched_error_copier(
+    original: _FunctionType, errs: _ErrorsHandled,
+) -> _FunctionType:
+    if inspect.iscoroutinefunction(original):
+        async def wrapper(self, *args, **kwargs):
+            original_result = await original(self, *args, **kwargs)
+            if id(self) in errs:
+                errs[id(original_result)] = original_result
+            return original_result
+    else:
+        def wrapper(self, *args, **kwargs):
+            original_result = original(self, *args, **kwargs)
+            if id(self) in errs:
+                errs[id(original_result)] = original_result
+            return original_result
+    return wraps(original)(wrapper)  # type: ignore
+
+
+_ERROR_HANDLING_PATCHERS: Final = {  # noqa: WPS407
```

Comment:
Do we need this to be mutable? Or will `types.MappingProxyType` work here?

Reply:
Nope. Chose the `dict` for simplicity. Can use a readonly proxy just fine

Reply:
Let's do this! We are kinda strict on mutability in this specific project 👍

Reply:
Naturally 😅 . It's a shame`MappingProxyType` isn't a full `frozendict`-like object, but we'll take it.